scripts/pylib/pytest-twister-ext/src/twister_ext/device/qemu_adapter.py

```diff
@@ -0,0 +1,196 @@
+        finally:
+            self._emulation_was_finished = True
+
+    def disconnect(self):
+        logger.debug('Closing connection')
+        self.connection.disconnect()
+
+    def stop(self) -> None:
+        """Stop device."""
+        time.sleep(0.1)  # give a time to end while loop in running simulation
+        if self._process is not None and self._process.returncode is None:
+            logger.debug('Stopping all running processes for PID %s', self._process.pid)
+            # kill all child subprocesses
+            for child in psutil.Process(self._process.pid).children(recursive=True):
+                try:
+                    os.kill(child.pid, signal.SIGTERM)
+                except ProcessLookupError:
+                    pass
+            # kill subprocess if it is still running
+            os.kill(self._process.pid, signal.SIGTERM)
+        if self._thread is not None:
+            self._thread.join(timeout=1)  # Should end immediately, but just in case we set timeout for 1 sec
+        if self._exc:
+            raise self._exc
+
+    def _wait_for_fifo(self):
+        for _ in range(int(self.booting_timeout_in_ms / 10) or 1):
+            if self.connection.is_open:
+                break
+            elif self._emulation_was_finished:
+                msg = 'Problem with starting QEMU'
+                logger.error(msg)
+                raise TwisterExtException(msg)
+            time.sleep(0.1)
+        else:
+            msg = 'Problem with starting QEMU - fifo file was not created yet'
+            logger.error(msg)
+            raise TwisterExtException(msg)
+
+    @property
+    def iter_stdout(self) -> Generator[str, None, None]:
+        if not self.connection:
+            return
+        # fifo file can be not create yet, so we need to wait for a while
+        self._wait_for_fifo()
+
+        # create unblocking reading from fifo file
+        q: Queue = Queue()
+
+        def read_lines():
+            while self.connection and self.connection.is_open:
+                try:
+                    line = self.connection.readline().decode('UTF-8').strip()
+                except (OSError, ValueError):
+                    # file could be closed already so we should stop reading
+                    break
+                if len(line) != 0:
+                    q.put(line)
+
+        t = threading.Thread(target=read_lines, daemon=True)
+        t.start()
+
+        end_time = time.time() + self.timeout
+        try:
+            while True:
+                try:
+                    stream = q.get(timeout=0.1)
+                    self.handler_log_file.handle(data=stream + '\n')
+                    yield stream
+                except Empty:  # timeout appeared
+                    pass
+                if time.time() > end_time:
+                    break
+        except KeyboardInterrupt:
+            # let thread to finish smoothly
+            pass
+        finally:
+            t.join(1)
```

Comment:
Another option would be to have QEMU connect its UARTs to Unix-domain sockets.  Here's a snippet I did for a test I was working on: https://gist.github.com/dreiss/63fed74aa7f43527a987916b646fc860 .  This scales to multiple UARTs and doesn't require a separate thread as long as you're draining the buffers frequently enough.